merkle/compact/tree.go

```diff
@@ -163,19 +163,20 @@ func (t *Tree) recalculateRoot(setNodeFn setNodeFunc) error {
 
 	index := t.size
 
-	var newRoot []byte
+	newRoot := make([]byte, t.hasher.Size())
 	first := true
 	mask := int64(1)
 	numBits := bits.Len64(uint64(t.size))
 	for bit := 0; bit < numBits; bit++ {
 		index >>= 1
 		if t.size&mask != 0 {
 			if first {
-				newRoot = t.nodes[bit]
+				copy(newRoot, t.nodes[bit])
 				first = false
 			} else {
-				newRoot = t.hasher.HashChildren(t.nodes[bit], newRoot)
-				if err := setNodeFn(bit+1, index, newRoot); err != nil {
+				newRoot := t.hasher.HashChildrenInto(t.nodes[bit], newRoot, newRoot)
```

Comment:
`HashChildrenInto` reads and writes to `newRoot` at the same time. Should it be specified in the interface's contract that such pattern is safe? (also applies below, and to other methods)

Reply:
Clarified docs.